### PR TITLE
array_type_field_filter: Add new contains filter to search on Array type field

### LINKED ISSUE #342: Malformed Array Literal
Hai,
I am using sequelize for the migration task instead of the default automigrate function. I have created a column "tags" with datatype Array in sequelize.  And created the model using "lb model" and defined the type of tags as the array of string. But when i POST to /api/users it throws malformed array literal error. Turns out the default automigrate function create the type of "tags" to "text" not to actual postgres array. Any way to use the default Postgres array type in here?
My model definition is as,

{
  "name": "user",
  "plural": "users",
  "base": "mfy-base",
  "idInjection": false,
  "options": {
    "validateUpsert": true
  },
  "forceId": false,
  "properties": {
    "id": {
      "type": "string",
      "id": true,
      "required": false,
      "defaultFn": "uuid",
      "postgresql": {
        "dataType": "uuid"
      }
    },
    "psid": {
      "type": "string"
    },
    "profile_pic": {
      "type": "string"
    },
    "gender": {
      "type": "string"
    },
    "locale": {
      "type": "string"
    },
    "timezone": {
      "type": "string"
    },
    "status": {
      "type": "string",
      "default": "unread"
    },
    "tags": {
      "type": ["string"]
    },
    "last_message": {
      "type": "object"
    },
    "last_fb_message_at": {
      "type": "date",
      "defaultFn": "now"
    },
    "mmail": {
      "type": "string"
    },
    "is_mmail_allowed": {
      "type": "boolean",
      "default": true
    },
    "is_messaging_allowed": {
      "type": "boolean",
      "default": true
    },
    "first_name": {
      "type": "string"
    },
    "last_name": {
      "type": "string"
    },
    "type": {
      "type": "string"
    }
  },
  "validations": [],
  "relations": {
    "page": {
      "type": "belongsTo",
      "model": "page",
      "foreignKey": "",
      "options": {
        "nestRemoting": true
      }
    },
    "events": {
      "type": "hasMany",
      "model": "event",
      "foreignKey": "",
      "options": {
        "nestRemoting": true
      }
    }
  },
  "acls": [],
  "methods": {}
}

### PR DESCRIPTION
**Pull request by shubhisood**
- Adds new filter named as contains.
- This filter will search on Array type fields in Postgres DB.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #342 

## Checklist

- [X] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [X] `npm test` passes on your machine
- [X] New tests added or existing tests modified to cover all changes
- [X] Code conforms with the [style guide](https://loopback.io/doc/en/contrib/style-guide-es6.html)
- [X] Commit messages are following our [guidelines](https://loopback.io/doc/en/contrib/git-commit-messages.html)

### CODE REVIEW

File: test/postgresql.test.js
Comment (shubhisood) [inline test/postgresql.test.js:267]: After adding this we do not need juggler related PR.
PFB PR Link:
https://github.com/strongloop/loopback-datasource-juggler/pull/1865

File: test/postgresql.test.js
Comment (bajtos) [inline test/postgresql.test.js:272]: I find this filter too complicated, would the following simpler version work too?

```js
Post.find({where: {categories: {'@>': ['AA']}}})
```

File: test/postgresql.test.js
Comment (bajtos) [inline test/postgresql.test.js:273]:
```suggestion
      if (err) return done(err);
```

That way the original error details and the stack trace is preserved.

File: test/postgresql.test.js
Comment (bajtos) [inline test/postgresql.test.js:266]: We prefer to use async/await style for all newly written tests. Can you please rewrite your test as follows?

```js
it('should support where filter for array type field', async () => {
  const post = await Post.find(/* your filter */);
  should.exist(post);
});
```

File: lib/postgresql.js
Comment (bajtos) [inline lib/postgresql.js:525]: I find the operator name `@>` as very cryptic. Have you considered using a more descriptive name like `contains`? 

Having said that, maybe the operator `@>` is something that PostgreSQL users are familiar with, and a new name like `contains` would just add learning overhead?

Let's discuss.

**Review by bajtos — COMMENTED** [review]
I like how it's possible to implement your desired feature inside the connector only :+1:

I have few comments on implementation details, PTAL above.

Please add a new section to README to explain what extended operators are supported (e.g. `@>`) and what is needed to allow the app to use them (set `allowExtendedOperators` in model and/or datasource settings).

File: lib/postgresql.js
Comment (bajtos) [reply lib/postgresql.js:525]: Another aspect to consider: this operator can be used in REST API clients too, e.g. to call 

```http
GET https://somehost/api/posts?filter={"where":{"category":{"@>":["AA"]}}}
```

From the point of view of REST API consumers (e.g. front-end developers), I think a descriptive operator name like `contains` is much better.

```http
GET https://somehost/api/posts?filter={"where":{"category":{"contains":["AA"]}}}
```

File: test/postgresql.test.js
Comment (shubhisood) [reply test/postgresql.test.js]: Yes, @bajtos it is working fine too.

File: test/postgresql.test.js
Comment (shubhisood) [reply test/postgresql.test.js]: Done with this test-case change

File: lib/postgresql.js
Comment (shubhisood) [reply lib/postgresql.js]: Done renamed operator from @> to contains

File: test/postgresql.test.js
Comment (shubhisood) [reply test/postgresql.test.js]: How we would be handling this with async await type test case? @bajtos 
I was thinking to write test case this way for handling failure of testcase with proper details.


it('should support where filter for array type field', async () => {
      Post.dataSource.settings.allowExtendedOperators = true;
      try {
        const post = await Post.find({where: {
          categories: {'contains': ['AA']},
        }});
        should.exist(post);
        post.length.should.equal(1);
      } catch (err) {
        should.not.exist(err);
      }
    });


Please tell me if it is correct, I will update current test case in this way only.

Comment (bajtos) [reply test/postgresql.test.js]: As a rule of thumb, the tests should pass any unexpected error to the test runner:
- Use `done(err)` when using callback
- Use regular error propagation when using async functions

```js
it('should support where filter for array type field', async () => {
      Post.dataSource.settings.allowExtendedOperators = true;
        const post = await Post.find({where: {
          categories: {'contains': ['AA']},
        }});
        should.exist(post);
        post.length.should.equal(1);
    });
```

**Review by bajtos — COMMENTED** [review]
The new version looks pretty good.

1. Please change the commit message to use the new operator name

2. Please add a new section to README to explain what extended operators are supported (e.g. `@>`) and what is needed to allow the app to use them (set `allowExtendedOperators` in model and/or datasource settings). Put yourself in the shoes of other LoopBack users - where would you look for information about `contains` operator and what info would you want to see to help you understand how to use this feature in your app?

**Comment by shubhisood** [comment]
@bajtos Thanks for review..
For Point#1 I have done the update.
For point number 2, I have following queries:
In which section of readme should I add this operator, As in current readme I couldn't find currently supported operators of postgres in this connector.

**Comment by bajtos** [comment]
> In which section of readme should I add this operator, As in current readme I couldn't find currently supported operators of postgres in this connector.

That's a good question. I believe the postgres connector so far supports only built-in LoopBack operators supported by all (or most?) connectors, so there was no need to explicitly list extended operators.

Now that you are implementing the first extended operator, I think you should simply add a new section. You can draw inspiration from MongoDB connector README:

- The section [Additional properties](https://github.com/strongloop/loopback-connector-mongodb/blob/4551cc0428a6954091981eb8e4861e7c1c9b37d8/README.md#additional-properties) mentions the configuration setting `allowExtendedOperators` and the extended operators unlocked by this option
- The section [Update operators](https://github.com/strongloop/loopback-connector-mongodb/blob/4551cc0428a6954091981eb8e4861e7c1c9b37d8/README.md#update-operators) adds more details

In the case of PostgreSQL and your new `contains` operator, we can either call the section something like `Query operators` or perhaps a more generic `Extended operators`.

**Comment by shubhisood** [comment]
> > In which section of readme should I add this operator, As in current readme I couldn't find currently supported operators of postgres in this connector.
> 
> That's a good question. I believe the postgres connector so far supports only built-in LoopBack operators supported by all (or most?) connectors, so there was no need to explicitly list extended operators.
> 
> Now that you are implementing the first extended operator, I think you should simply add a new section. You can draw inspiration from MongoDB connector README:
> 
> * The section [Additional properties](https://github.com/strongloop/loopback-connector-mongodb/blob/4551cc0428a6954091981eb8e4861e7c1c9b37d8/README.md#additional-properties) mentions the configuration setting `allowExtendedOperators` and the extended operators unlocked by this option
> * The section [Update operators](https://github.com/strongloop/loopback-connector-mongodb/blob/4551cc0428a6954091981eb8e4861e7c1c9b37d8/README.md#update-operators) adds more details
> 
> In the case of PostgreSQL and your new `contains` operator, we can either call the section something like `Query operators` or perhaps a more generic `Extended operators`.

Added the readme section for the new contains operator

**Comment by bajtos** [comment]
Thank you @shubhisood for the update. I pushed few more commits to fix the failing builds and improve README content. Let's wait for CI results before landing.

**Review by bajtos — APPROVED** [review]

**Comment by bajtos** [comment]
@slnode test please

**Comment by bajtos** [comment]
@slnode test please

**Comment by bajtos** [comment]
Landed, thank you @shubhisood for the contribution ❤️

**Comment by shubhisood** [comment]
> Landed, thank you @shubhisood for the contribution

Thanks for merging @bajtos

**Pull request merged**
